### ISSUE TITLE
How to use the options variable (opts)

### ISSUE THREAD

**Issue by JanFrederickUnnewehr** [PyPSA/pypsa-eur#37]
Hello,
I have tried to figure out which possibilities the different "opts" variables have in [config.yaml](https://github.com/PyPSA/pypsa-eur/blob/master/config.yaml). 
I didn't find any information about the different "opts" that are listed in the file. What do the different options?

In the documentation, another option is listed after the "opts" variable. The "all countries" option. Has this option been replaced by the countries the [config.yaml](https://github.com/PyPSA/pypsa-eur/blob/master/config.yaml)?

Thanks in advance,
Best Freddy

**Comment by coroa** [comment]
That's definitely a gap in the documentation, I agree!


## `opts` wildcard

The `opts` wildcard triggers optional constraints, which are activated in either `prepare_network` (here: or the `solve_network` step. It may hold multiple triggers separated by `-`, i.e. `Co2L-3H` contains the `Co2L` trigger and the `3H` switch. There are currently:

Trigger | Description | Definition | Status
----------|----------------|-------------|----------
`Co2L`    | Add an overall Co2 emissions limit configured in `electricity: co2limit` | `prepare_network`: [`add_co2limit()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L19) and its [callee](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L154) | In active use
`nH`, i.e. `2H`-`6H` | Resample the time-resolution by averaging over every `n` snapshots | `prepare_network`: [`average_every_nhours()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L110) and its [callee](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L146) | In active use
`Ep` | Add cost for a Co2 price configured in `costs: emission_prices: co2` to `marginal_cost` of generators | `prepare_network`: [`add_emission_prices()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L24) and its [callee](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L158) | Broken, commented out
`BAU` | Add a per-`carrier` minimal overall capacity, i.e. at least `40GW` of `OCGT` in Europe, configured in `electricity: BAU_mincapacities` | `solve_network`: [`add_opts_constraints()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/solve_network.py#L66) | Was used successfully in [`pypsa-za`](https://github.com/PyPSA/pypsa-za) sister model. Should be renamed.
`SAFE` | Add a capacity reserve margin of a certain fraction above the peak demand, to which renewable generators and storage do *not* contribute. Ignores network. | `solve_network` [`add_opts_constraints()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/solve_network.py#L73) | Was used in `pypsa-za` as well. Untested in `pypsa-eur`.

## `country` wildcard

The rules `make_summary` (generating summaries of all or a subselection of the solved networks) and `plot_p_nom_max` (for plotting the cumulative generation potentials for renewable technologies) can be narrowed to individual countries using the `country` wildcard.

If `country = all`, then the rule acts on the network for all countries defined in `config.yaml`. If otherwise `country = DE` or another country code, then the network is narrowed to buses of this country for the rule. For example to get a summary of the energy generated in Germany (in the solution for Europe) use:
```
snakemake results/summaries/elec_s_all_lall_Co2L-3H_DE
```

## TODO

Please prepare a pull request to add this description to `README.md`!

**Comment by JanFrederickUnnewehr** [comment]
Hey Jonas,
thanks for the good explanation. This is really helpful for me. Should I do the pull request or do I need special permissions to do that?
Best Freddy

**Comment by coroa** [comment]
Hi Jan,

yes, i am happy, if you submit the PR. No permissions are needed for that as long as you push changes to an own fork of the repository; for example as described in the gist [Github-Forking]https://gist.github.com/Chaser324/ce0505fbed06b947d962).

Best

**Comment by coroa** [comment]
I'm unsure whether you can see the Markdown code in my comments thus a copy of the code from before

```
## `opts` wildcard

The `opts` wildcard triggers optional constraints, which are activated in either `prepare_network` or the `solve_network` step. It may hold multiple triggers separated by `-`, i.e. `Co2L-3H` contains the `Co2L` trigger and the `3H` switch. There are currently:

Trigger | Description | Definition | Status
----------|----------------|-------------|----------
`Co2L`    | Add an overall Co2 emissions limit configured in `electricity: co2limit` | `prepare_network`: [`add_co2limit()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L19) and its [callee](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L154) | In active use
`nH`, i.e. `2H`-`6H` | Resample the time-resolution by averaging over every `n` snapshots | `prepare_network`: [`average_every_nhours()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L110) and its [callee](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L146) | In active use
`Ep` | Add cost for a Co2 price configured in `costs: emission_prices: co2` to `marginal_cost` of generators | `prepare_network`: [`add_emission_prices()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L24) and its [callee](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/prepare_network.py#L158) | Broken, commented out
`BAU` | Add a per-`carrier` minimal overall capacity, i.e. at least `40GW` of `OCGT` in Europe, configured in `electricity: BAU_mincapacities` | `solve_network`: [`add_opts_constraints()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/solve_network.py#L66) | Was used successfully in [`pypsa-za`](https://github.com/PyPSA/pypsa-za) sister model. Should be renamed.
`SAFE` | Add a capacity reserve margin of a certain fraction above the peak demand, to which renewable generators and storage do *not* contribute. Ignores network. | `solve_network` [`add_opts_constraints()`](https://github.com/PyPSA/pypsa-eur/blob/6b964540ed39d44079cdabddee8333f486d0cd63/scripts/solve_network.py#L73) | Was used in `pypsa-za` as well. Untested in `pypsa-eur`.

## `country` wildcard

The rules `make_summary` (generating summaries of all or a subselection of the solved networks) and `plot_p_nom_max` (for plotting the cumulative generation potentials for renewable technologies) can be narrowed to individual countries using the `country` wildcard.

If `country = all`, then the rule acts on the network for all countries defined in `config.yaml`. If otherwise `country = DE` or another country code, then the network is narrowed to buses of this country for the rule. For example to get a summary of the energy generated in Germany (in the solution for Europe) use:
snakemake results/summaries/elec_s_all_lall_Co2L-3H_DE
```

**Comment by JanFrederickUnnewehr** [comment]
Hey Jonas, I will add you comments to the `README.md`. I can't see the markdown code from your posts, so thanks for the text. I will use it for the `README.md` file.

**Comment by fneum** [comment]
Related to #4

**Comment by fneum** [comment]
Since you asked for it @JanFrederickUnnewehr, there's now a [documentation branch](https://github.com/PyPSA/PyPSA-Eur/tree/documentation) with a basic outline for a documentation (362f183) for which PRs can be prepared. These PRs can be incremental (e.g. about a specific rule). Once we have a good first issue we can pull this into the master branch.

Workflow: Fork this repository, checkout the branch `documentation`, make improvements, submit PR to branch `documentation`.

**Comment by fneum** [comment]
closed in #65
